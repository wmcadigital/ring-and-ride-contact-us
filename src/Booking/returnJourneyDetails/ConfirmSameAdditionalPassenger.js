import { useEffect } from "react";
import PropTypes from "prop-types";
import { useFormState, useForm } from "react-final-form";

import FormSection from "../../common/FormSection";
import RadioGroup from "../../common/RadioGroup";
import Question from "../../common/Question";
import RadioButton from "../../common/RadioButton";
import FieldError from "../../common/FieldError";
import ProgressIndicator from "../../common/ProgressIndicator";
import getSectionPositionInfo from "../getSectionPosition";
import getAboutReturnJourneySectionName from "./getAboutReturnJourneySectionName";

import { required } from "../../common/validation";

const ConfirmSameAdditionalPassenger = ({
  setConfirmSameAdditionalPassenger,
}) => {
  const stateApi = useFormState();
  const formApi = useForm();

  const formValues = stateApi.values;
  const bookingParty = formValues["bookingParty"];

  const error = stateApi.submitFailed
    ? stateApi.errors?.confirmSameAdditionalPassenger
    : null;

  const question =
    bookingParty === "mySelf"
      ? "Are the same number of people joining you on the return journey?"
      : `Are the same number of people joining ${formValues["firstName"]} on the return journey?`;

  const aloneLabel =
    bookingParty === "mySelf"
      ? "I'm returning alone"
      : "They are returning alone";

  useEffect(() => {
    if (formValues.confirmSameAdditionalPassenger) {
      setConfirmSameAdditionalPassenger(
        formValues.confirmSameAdditionalPassenger
      );
    }
  }, [formValues.confirmSameAdditionalPassenger]);

  const radioButtonClick = (alone) => {
    const passengerNumbersReturn = alone
      ? 0
      : formValues.additionalPassengerNumbers;
    formApi.mutators.setFormAttribute(
      "additionalReturnPassengerNumbers",
      passengerNumbersReturn
    );
  };

  return (
    <FormSection>
      <ProgressIndicator
        sectionPosition={getSectionPositionInfo("3", stateApi)}
        sectionName={getAboutReturnJourneySectionName(stateApi)}
      />
      <Question text={question} />
      <RadioGroup error={error}>
        <FieldError text={error} />
        <RadioButton
          key={1}
          label="Yes"
          validation={required}
          value="yes"
          fieldName="confirmSameAdditionalPassenger"
          onClickCallback={() => radioButtonClick(false)}
        />
        <RadioButton
          key={2}
          label="No"
          validation={required}
          value="no"
          fieldName="confirmSameAdditionalPassenger"
          onClickCallback={() => radioButtonClick(false)}
        />
        <RadioButton
          key={3}
          label={aloneLabel}
          validation={required}
          value="alone"
          fieldName="confirmSameAdditionalPassenger"
          onClickCallback={() => radioButtonClick(true)}
        />
      </RadioGroup>
    </FormSection>
  );
};

ConfirmSameAdditionalPassenger.propTypes = {
  setConfirmSameAdditionalPassenger: PropTypes.func,
};

ConfirmSameAdditionalPassenger.defaultProps = {
  setConfirmSameAdditionalPassenger: () => {},
};

export default ConfirmSameAdditionalPassenger;
