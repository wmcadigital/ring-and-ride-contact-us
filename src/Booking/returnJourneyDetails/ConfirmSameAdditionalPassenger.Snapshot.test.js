import { useFormState } from "react-final-form";
import { createRenderer } from "react-test-renderer/shallow";

import ConfirmSameAdditionalPassenger from "./ConfirmSameAdditionalPassenger";

jest.mock("react-final-form");

describe("ConfirmSameAdditionalPassenger - snapshots", () => {
  it("individual is filling out form for themselves", () => {
    useFormState.mockImplementationOnce(() => ({
      values: {
        bookingParty: "mySelf",
      },
    }));
    const renderer = createRenderer();

    renderer.render(<ConfirmSameAdditionalPassenger />);

    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  it("individual is filling out form for someone else", () => {
    useFormState.mockImplementationOnce(() => ({
      values: {
        bookingParty: "behalfSomeone",
        firstName: "John",
      },
    }));
    const renderer = createRenderer();

    renderer.render(<ConfirmSameAdditionalPassenger />);

    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  it("renders with an error", () => {
    useFormState.mockImplementationOnce(() => ({
      values: {
        bookingParty: "mySelf",
        firstName: "John",
      },
      submitFailed: true,
      errors: { confirmSameAdditionalPassenger: "Required" },
    }));
    const renderer = createRenderer();

    renderer.render(<ConfirmSameAdditionalPassenger />);

    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });
});
