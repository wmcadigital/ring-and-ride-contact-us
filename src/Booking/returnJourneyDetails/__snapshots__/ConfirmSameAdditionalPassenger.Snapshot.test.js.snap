// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmSameAdditionalPassenger - snapshots individual is filling out form for someone else 1`] = `
<FormSection>
  <ProgressIndicator
    sectionName="Plan their return journey"
    sectionPosition="Section 3 of 4"
  />
  <Question
    text="Are the same number of people joining John on the return journey?"
  />
  <RadioGroup
    error={null}
  >
    <FieldError
      text={null}
    />
    <RadioButton
      fieldName="confirmSameAdditionalPassenger"
      label="Yes"
      onClickCallback={[Function]}
      validation={[Function]}
      value="yes"
    />
    <RadioButton
      fieldName="confirmSameAdditionalPassenger"
      label="No"
      onClickCallback={[Function]}
      validation={[Function]}
      value="no"
    />
    <RadioButton
      fieldName="confirmSameAdditionalPassenger"
      label="They are returning alone"
      onClickCallback={[Function]}
      validation={[Function]}
      value="alone"
    />
  </RadioGroup>
</FormSection>
`;

exports[`ConfirmSameAdditionalPassenger - snapshots individual is filling out form for themselves 1`] = `
<FormSection>
  <ProgressIndicator
    sectionName="Plan your return journey"
    sectionPosition="Section 3 of 4"
  />
  <Question
    text="Are the same number of people joining you on the return journey?"
  />
  <RadioGroup
    error={null}
  >
    <FieldError
      text={null}
    />
    <RadioButton
      fieldName="confirmSameAdditionalPassenger"
      label="Yes"
      onClickCallback={[Function]}
      validation={[Function]}
      value="yes"
    />
    <RadioButton
      fieldName="confirmSameAdditionalPassenger"
      label="No"
      onClickCallback={[Function]}
      validation={[Function]}
      value="no"
    />
    <RadioButton
      fieldName="confirmSameAdditionalPassenger"
      label="I'm returning alone"
      onClickCallback={[Function]}
      validation={[Function]}
      value="alone"
    />
  </RadioGroup>
</FormSection>
`;

exports[`ConfirmSameAdditionalPassenger - snapshots renders with an error 1`] = `
<FormSection>
  <ProgressIndicator
    sectionName="Plan your return journey"
    sectionPosition="Section 3 of 4"
  />
  <Question
    text="Are the same number of people joining you on the return journey?"
  />
  <RadioGroup
    error="Required"
  >
    <FieldError
      text="Required"
    />
    <RadioButton
      fieldName="confirmSameAdditionalPassenger"
      label="Yes"
      onClickCallback={[Function]}
      validation={[Function]}
      value="yes"
    />
    <RadioButton
      fieldName="confirmSameAdditionalPassenger"
      label="No"
      onClickCallback={[Function]}
      validation={[Function]}
      value="no"
    />
    <RadioButton
      fieldName="confirmSameAdditionalPassenger"
      label="I'm returning alone"
      onClickCallback={[Function]}
      validation={[Function]}
      value="alone"
    />
  </RadioGroup>
</FormSection>
`;
